include/ginkgo/core/config/property_tree.hpp

```diff
@@ -0,0 +1,210 @@
+    using map_type = std::map<key_type, pnode>;
+    using array_type = std::vector<pnode>;
+
+    /**
+     * status_t is the indicator for the current node storage.
+     */
+    enum class status_t { empty, array, boolean, real, integer, string, map };
+
+    /**
+     * Default constructor: create an empty node
+     */
+    pnode();
+
+    /**
+     * Constructor for bool
+     *
+     * @param boolean  the bool type value
+     */
+    pnode(bool boolean);
+
+    /**
+     * Constructor for integer with all integer type
+     *
+     * @tparam T  input type
+     *
+     * @param integer  the integer type value
+     */
+    template <typename T,
+              std::enable_if_t<std::is_integral<T>::value>* = nullptr>
+    pnode(T integer);
+
+    /**
+     * Constructor for string
+     *
+     * @param str  string type value
+     */
+    pnode(const std::string& str);
+
+    /**
+     * Constructor for char* (otherwise, it will use bool)
+     *
+     * @param str  the string like "..."
+     */
+    pnode(const char* str);
+
+    /**
+     * Constructor for double (and also float)
+     *
+     * @param real  the floating point type value
+     */
+    pnode(double real);
+
+    /**
+     * Constructor for array
+     *
+     * @param array  an pnode array
+     */
+    pnode(const array_type& array);
+
+    /**
+     * Constructor for map
+     *
+     * @param map  a (string, pnode)-map
+     */
+    pnode(const map_type& map);
+
+    /**
+     * bool conversion. It's true if and only if it is not empty.
+     */
+    operator bool() const noexcept;
+
+    /**
+     * Get the current node status.
+     *
+     * @return the status
+     */
+    status_t get_status() const;
+
+    /**
+     * Get the array. It will throw error if the current node does not hold an
+     * array.
+     *
+     * @return the array const reference
+     */
+    const array_type& get_array() const;
+
+    /**
+     * Get the map. It will throw error if the current node does not hold an
+     * map.
+     *
+     * @return the map const reference
+     */
+    const map_type& get_map() const;
+
+    /**
+     * Get the boolean value. It will throw error if the current node does not
+     * hold an boolean value.
+     *
+     * @return the boolean value
+     */
+    bool get_boolean() const;
+
+    /**
+     * Get the integer value. It will throw error if the current node does not
+     * hold an integer value.
+     *
+     * @return the integer value with type int64_t
+     */
+    std::int64_t get_integer() const;
+
+    /**
+     * Get the real floating point value. It will throw error if the current
+     * node does not hold an real value.
+     *
+     * @return the real value with type double
+     */
+    double get_real() const;
+
+    /**
+     * Get the string. It will throw error if the current node does not hold an
+     * string.
+     *
+     * @return the string
+     */
+    std::string get_string() const;
```

Comment:
I think it is cheap here, so I put this like the other bool, double, int.
Or, do you think they all should be returned const reference?

Reply:
This one has not been addressed yet

Reply:
I think I prefer just return it as value as the other integer/bool now. If that user changes the value from the return does not change the property tree itself, it is what user wants to do in my mind.

Reply:
> it is what user wants to do in my mind

This operation does not make any sense though, since it creates a temporary copy, clears it and then does not give the user any way to access it. I can see users making the assumption that you can modify the object that way and being surprised - a compiler error is a clear message that it's not possible.
The same is also not possible with the return values from the primitive value access functions (e.g. `get_integer()++` fails to compile).

Reply:
That makes sense.
I originally thought the case `auto a = get_string();`. I thought it will be const std::string&, which is inconsistent with the others. 
However, type of `a` is actually std::string unless using `auto& a = get_string()`